tests/MongoDB.Driver.Core.Tests/Core/Connections/ConnectionInitializerTests.cs

```diff
@@ -67,6 +69,63 @@ public void InitializeConnection_should_acquire_connectionId_from_isMaster_respo
             result.ConnectionId.ServerValue.Should().Be(1);
         }
 
+        [Theory]
+        [ParameterAttributeData]
+        public void InitializeConnection_should_speculatively_authenticate_with_SCRAM_SHA(
+            [Values("default", "SCRAM-SHA-256", "SCRAM-SHA-1")] string authenticatorType,
+            [Values(false, true)] bool async)
+        {
+            var isMasterReply = MessageHelper.BuildReply(
+                RawBsonDocumentHelper.FromJson("{ ok : 1, connectionId : 1 }"));
+            var buildInfoReply = MessageHelper.BuildReply(
+                RawBsonDocumentHelper.FromJson("{ ok : 1, version : \"4.2.0\" }"));
+            var credentials = new UsernamePasswordCredential(
+                source: "Voyager", username: "Seven of Nine", password: "Omega-Phi-9-3");
+            var authenticator = CreateAuthenticator();
+            var connectionSettings = new ConnectionSettings(new[] { authenticator });
+            var connection = new MockConnection(__serverId, connectionSettings, eventSubscriber: null);
+            connection.EnqueueReplyMessage(isMasterReply);
+            connection.EnqueueReplyMessage(buildInfoReply);
+
+            // We expect authentication to fail since we have not enqueued the expected authentication replies
+            try
+            {
+                if (async)
+                {
+                    _subject.InitializeConnectionAsync(connection, CancellationToken.None).GetAwaiter().GetResult();
+                }
+                else
+                {
+                     _subject.InitializeConnection(connection, CancellationToken.None);
+                }
+            }
+            catch (InvalidOperationException ex)
+            {
+                ex.Message.Should().Be("Queue empty.");
```

Comment:
Essentially, this is a white-box test that relies on the fact that `CreateInitialIsMasterCommand` is called before the exception is thrown. Alternatively, we could use reflection to test `CreateInitialIsMasterCommand` instead of only partially populating the MockConnection queue and catching the exception that occurs when we try to authenticate with no messages left on the queue. Perhaps a minor advantage of testing it this way is that we also ensure that InitializeConnection definitely calls `CreateInitialIsMasterCommand` as this test will fail if that ever changes.

Reply:
Sorry, I didn't understand the full idea originally. 
Now If I'm not mistaken the idea of this test is just asserting that `#X#Authenticator.CustomizeInitialIsMasterCommand` adds `speculativeAuthenticate` node, maybe we will just assert calling of this method and his result?
This test seems unclear a bit..
I will rely on your and other reviewers opinion here
This seems unnecessary difficult.

Reply:
Normally we would unit test `CreateInitialIsMasterCommand` if the intent is to test `CreateInitialIsMasterCommand`.

The tests for `InitializeConnection` would normally be different. For example, it `InitializeConnection` do what it was supposed to do (assuming `CreateInitialIsMasterCommand` is correct).

Reply:
I've added a unit test but mostly left this test as is it is because I still feel like it doesn't do any harm to have this test. I've renamed it to better express my intent. If everyone insists that this test needs to go, I'll remove it. 

IMHO: The test might not be perfect, but I don't think it's harmful either. It's... "good": adds value, but not the ideal implementation, but good enough to get the job done.

If reviewers disagree, I'm happy to remove this test or update this test to the "perfect" version, which I think would be to use a Moq to ensure that `InitializeConnection` calls an `IAuthenticator`'s `CustomizeInitialIsMasterCommand`

Reply:
I will rely on @rstam's opinion here :)

Reply:
So you are testing that `InitializeConnection` called `CustomizeInitializeIsMasterCommand` not by using mocks and verifying that the call was made but instead by observing the side effects that could only have occured if `CustomizeInitializeIsMasterCommand` was actually called?

I think that's OK. Setting up mocks to do this might be really hard but I'm not sure.